controllers/awscluster_controller_test.go

```diff
@@ -13,42 +13,1154 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	"github.com/aws/aws-sdk-go/service/elb"
+	"github.com/golang/mock/gomock"
 	. "github.com/onsi/gomega"
-	"golang.org/x/net/context"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	ctrl "sigs.k8s.io/controller-runtime"
-	"sigs.k8s.io/controller-runtime/pkg/client"
+	"k8s.io/client-go/tools/record"
 
 	infrav1 "sigs.k8s.io/cluster-api-provider-aws/api/v1beta1"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/scope"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/services"
+	ec2Service "sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/services/ec2"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/services/ec2/mock_ec2iface"
+	elbService "sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/services/elb"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/services/elb/mock_elbiface"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/services/network"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/services/securitygroup"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/util"
 )
 
-func TestAWSClusterReconciler(t *testing.T) {
-	g := NewWithT(t)
-	ctx := context.Background()
+func TestAWSClusterReconciler_IntegrationTests(t *testing.T) {
+	var (
+		reconciler AWSClusterReconciler
+		mockCtrl   *gomock.Controller
+		recorder   *record.FakeRecorder
+	)
+	var secret = &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-secret",
+			Namespace: "capa-system",
+		},
+		Data: map[string][]byte{
+			"AccessKeyID":     []byte("access-key-id"),
+			"SecretAccessKey": []byte("secret-access-key"),
+			"SessionToken":    []byte("session-token"),
+		},
+	}
+
+	setup := func(t *testing.T, g *WithT) {
+		t.Helper()
+		mockCtrl = gomock.NewController(t)
+		recorder = record.NewFakeRecorder(10)
+		reconciler = AWSClusterReconciler{
+			Client:   testEnv.Client,
+			Recorder: recorder,
+		}
+	}
 
-	reconciler := &AWSClusterReconciler{
-		Client: testEnv.Client,
+	teardown := func(t *testing.T, g *WithT) {
+		t.Helper()
+		mockCtrl.Finish()
 	}
 
-	instance := &infrav1.AWSCluster{ObjectMeta: metav1.ObjectMeta{Name: "foo", Namespace: "default"}}
-	instance.Default()
+	t.Run("Should not create VPC in case of unmanaged VPC", func(t *testing.T) {
+		g := NewWithT(t)
+		mockCtrl = gomock.NewController(t)
+		ec2Mock := mock_ec2iface.NewMockEC2API(mockCtrl)
+		elbMock := mock_elbiface.NewMockELBAPI(mockCtrl)
+		expect := func(m *mock_ec2iface.MockEC2APIMockRecorder, e *mock_elbiface.MockELBAPIMockRecorder) {
+			mockedCreateVPCCalls(m)
+			mockedCreateSGCalls(m)
+			mockedCreateLBCalls(e)
+			mockedDescribeInstanceCall(m)
+		}
+		expect(ec2Mock.EXPECT(), elbMock.EXPECT())
+
+		capaNS := &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "capa-system",
+			},
+		}
+		g.Expect(testEnv.Create(ctx, capaNS)).To(Succeed())
+
+		controllerIdentity := createControllerIdentity(g)
+		ns, err := testEnv.CreateNamespace(ctx, fmt.Sprintf("integ-test-%s", util.RandomString(5)))
```

Comment:
No need to create a random name ourself as it is already handled by GenerateName. 
Ref: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#idempotency 
Suggestion: Can we consider using this for naming other resources like AWSCluster during our integration tests. This will safeguard us against name collision.

Reply:
here we need random naming otherwise the cleanup sometimes does not happen properly.
I can add random name for AWSCluster in integ tests

Reply:
Cleanup should not be an issue as API server returns the object with the 'Name' field set. We can use this name during cleanup. This will also solve your above concern for using GenerateName with test objects like AWSCluster.   GenerateName helps to solve the problem of maintaining uniqueness of object names without using any custom naming technique in our tests.

Reply:
@Madhur97 I didn't get you point here

Reply:
what i meant with cleanup is, t.Cleanup sometimes doesn't clear the objects, that's  why we used random namespaces, and we cannot use random name for AWSCluster as you asked, since we face problem during mocking as mentioned here:

> We can't add random name for AWSCluster in integ tests, as some of the mocking call tags contain AWSCluster name so we won't know what to return as the AWSCluster name is random, here is an example. That's why we are using random namespace.

Reply:
I am suggesting to remove custom code for random name generation for namespaces and rely on 'GenerateName' to do that for you. 
Creating random names for other objects like AWSCluster was just a suggestion, but I guess we do not need to that if namespaces are already different.